### ISSUE TITLE
Low loss, but bad prediction

### ISSUE THREAD

**Issue by rodrigo2019** [experiencor/basic-yolo-keras#7]
Hello, I'm trying to train in my own dataset, when I load a pre trained weights I get low loss function, around 0.2, but if I do not use I get something like 6 or 7. But even using the pre trained weights and trying to predict the train dataset I dont get any correct prediction. I tried disable the augmentation data for overfitting the train dataset ( for all data and just for a single image as well), but did not worked, also I tried changing the threshold
Also I modified the data generator and the parser for reading data from txt file.
the reference of this adaptation is from here: https://github.com/yhenon/keras-frcnn
Link:[yolo_lib.zip](https://github.com/experiencor/basic-yolo-keras/files/1270107/yolo_lib.zip)

The faster RCNN looks to be more easy to train, even having a high cost per epoch, but corverge more quickly than YOLO

ps: My own dataset have around 1k images with around 7k objects.

Thank you for sharing your code, it is very easy to read and understand what you did.

**Comment by experiencor** [comment]
I initially thought that you might have overfitted the training set due to the lack of validation set in my implementation but you also said the network failed to overfit even a single image. This is very wrong. Can I have the single image that you used to overfit the network? I want to reproduce what's going on.

**Comment by rodrigo2019** [comment]
the .txt file work in this way: File_path,X1,Y1,X2,Y2,CLASS
I tried to set all classes in a single class (character class), replacing class name for "CHAR_CLASS"

[Sample Data.zip](https://github.com/experiencor/basic-yolo-keras/files/1272484/Sample.Data.zip)

I thought that my modification was wrong, but I checked twice my data generator, and the generator make the same pattern as yours, that is: (sample,grid_y,grid_x,boxes, class probabilities + position (4+1+classes)), for example : (8,13,13,5,25) for y_batch with 8 samples, 20 classes and 5 anchors.

**Comment by experiencor** [comment]
I was able to overfit the 10 images that you set me. I initially trained the network with base learning rates of 1e-3, 1e-4, then 1e-5 - each with 100 epoch. The batch size is 10. The final loss is 0.0885. Some sample predictions are shown below:

![download](https://user-images.githubusercontent.com/5645503/30003189-481f77c4-90ec-11e7-8f42-62356b4e3c05.png)

![download 1](https://user-images.githubusercontent.com/5645503/30003190-4bc614b4-90ec-11e7-94b5-aa5ddbda5494.png)

![download 2](https://user-images.githubusercontent.com/5645503/30003191-4ee5c856-90ec-11e7-9790-de76086b2e3a.png)

You may try to:

+ Use only one class (CLASS = 1) and one label (LABELS = ['CHAR_CLASS']). So the network will only predict one class CHAR_CLASS.

+ Follow the learning schedule tried by me, 1e-3 => 1e-4 => 1e-5. The aim is to get the loss around 0.08 for reasonable prediction.

Please let me know if it can help.

**Comment by rodrigo2019** [comment]
Thank you a lot, the problem was the high Loss, 0.2 is not low enough. I did not had luck with scheduling stochastic gradient, the best loss was around 1, but with adam I got 0.0125 after 250 epochs.

![image](https://user-images.githubusercontent.com/8016520/30010197-633e7786-9104-11e7-9c50-16a15712183f.png)

But I think I'm doing some mistake after edting your code, because I got some correct predictions, but theses predictions have a offset in Y
![notok](https://user-images.githubusercontent.com/8016520/30010225-aa57fe1c-9104-11e7-978a-c9194fcf4dd3.jpg)
If I change this lines in the interpreter:
**xmin  = int((box.x - box.w/2) * image.shape[1])
xmax  = int((box.x + box.w/2) * image.shape[1])
ymin  = int((box.y - box.h/2) * image.shape[0])
ymax  = int((box.y + box.h/2) * image.shape[0])**

to this:

**xmin  = int((box.x - box.w/2) * image.shape[1])
xmax  = int((box.x + box.w/2) * image.shape[1])
ymin  = int((box.y - box.h) * image.shape[0])
ymax  = int((box.y ) * image.shape[0])**

the predictions looks correct
![ok](https://user-images.githubusercontent.com/8016520/30010286-1172332e-9105-11e7-8ebd-a1d197783e1e.jpg)

First I thought the problem was when I forgot to divide the image by 255.0, because I commented out all your augmentation method and forgot to keep that division, but if I dont divide the image by 255 when I test the network, the network work correctly in the same way

Could you tell me how did you read my data.txt?

ps: that green line going to right side it is a bug with opencv 3.2 + python 3.5, some times that lines occours.

**Comment by experiencor** [comment]
This is my code to read your datat.txt

``
all_imgs_char = []
last_filename = ''

for line in open('Sample/data.txt'):
    filename, xmin, ymin, xmax, ymax, label = [string.strip() for string in line.split(',')]
    filename = 'Sample/' + filename.replace('\\', '/')
    
    if filename != last_filename:
        img = cv2.imread(filename)
        one_img = {'filename': filename,
                   'height': img.shape[0],
                   'width': img.shape[1],
                   'object': [{
                       'name': 'CHAR_CLASS',
                       'xmin': int(xmin),
                       'ymin': int(ymin),
                       'xmax': int(xmax),
                       'ymax': int(ymax)
                   }]}
        
        last_filename = filename
        all_imgs_char.append(one_img)
    else:
        all_imgs_char[-1]['object'].append({
           'name': 'CHAR_CLASS',
           'xmin': int(xmin),
           'ymin': int(ymin),
           'xmax': int(xmax),
           'ymax': int(ymax)        
        })
        

print all_imgs_char
``

**Comment by rodrigo2019** [comment]
I found my mistake, everything is working now, thank you.
